src/Kestrel/ServerRequest.cs

```diff
@@ -231,7 +233,53 @@ bool IHttpUpgradeFeature.IsUpgradableRequest
                 }
             }
             _frame.ProduceStart();
-            return _frame.DuplexStream;
+            return Task.FromResult<Stream>(_frame.DuplexStream);
+        }
+
+        IPAddress IHttpConnectionFeature.RemoteIpAddress
+        {
+            get
+            {
+                IPAddress ip;
+                var result = IPAddress.TryParse(_frame.RemoteIpAddress, out ip);
+                return result ? ip : IPAddress.Any;
+            }
+            set
+            {
+                throw new NotSupportedException();
```

Comment:
I think we're talking past each other.  The I*Feature interfaces represent accessing and storing per-request data for the web app.  While the data (like paths, IPs, etc.) initially come from underlying sources (Frame, Connection, etc.), the web application needs to be able to update all of the fields while processing the request. These updates do not effect the original data source, they are only for use by the higher level application.

For example: My server is hosted behind a reverse proxy. The RemoteIpAddress as reported by Kestrel is always the IP of the reverse proxy.  However, the proxy forwards me the real remote IP in an x-forwarded header. I've written a middleware that will take the value from the x-forwarded header and set it to the RemoteIpAddress field. This way only my middleware needs to be aware of the x-forwarded header, while the rest of the application can just look at the RemoteIpAddress field.  The underlying Connection is never modified, ServerRequest just stores the modified value internally for the duration of the request.